sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/Cast.java

```diff
@@ -0,0 +1,582 @@
+                  }
+                }))
+        .setRowSchema(outputSchema());
+  }
+
+  @VisibleForTesting
+  static Row castRow(Row input, Schema inputSchema, Schema outputSchema) {
+    Row.Builder output = Row.withSchema(outputSchema);
+    for (int i = 0; i < outputSchema.getFieldCount(); i++) {
+      Schema.Field outputField = outputSchema.getField(i);
+
+      int fromFieldIdx = inputSchema.indexOf(outputField.getName());
+      Schema.Field inputField = inputSchema.getField(fromFieldIdx);
+
+      Object inputValue = input.getValue(fromFieldIdx);
+
+      if (inputValue == null) {
+        output.addValue(null);
+      } else {
+        Object value = castObject(inputValue, inputField.getType(), outputField.getType());
+        output.addValue(value);
+      }
+    }
+
+    return output.build();
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("unchecked")
+  static Object castObject(
+      Object inputValue, Schema.FieldType inputFieldType, Schema.FieldType outputFieldType) {
+
+    if (inputFieldType.getTypeName() == ROW) {
+      return castRow(
+          (Row) inputValue, inputFieldType.getRowSchema(), outputFieldType.getRowSchema());
+    } else if (inputFieldType.getTypeName() == ARRAY) {
+      List<Object> inputValues = (List<Object>) inputValue;
+      List<Object> outputValues = new ArrayList<>(inputValues.size());
+
+      for (Object elem : inputValues) {
+        outputValues.add(
+            castObject(
+                elem,
+                inputFieldType.getCollectionElementType(),
+                outputFieldType.getCollectionElementType()));
+      }
+
+      return outputValues;
+    } else if (inputFieldType.getTypeName() == MAP) {
+      Map<Object, Object> inputMap = (Map<Object, Object>) inputValue;
+      Map<Object, Object> outputMap = Maps.newHashMapWithExpectedSize(inputMap.size());
+
+      for (Map.Entry<Object, Object> entry : inputMap.entrySet()) {
+        Object outputKey =
+            castObject(
+                entry.getKey(), inputFieldType.getMapKeyType(), outputFieldType.getMapKeyType());
+        Object outputValue =
+            castObject(
+                entry.getValue(),
+                inputFieldType.getMapValueType(),
+                outputFieldType.getMapValueType());
+
+        outputMap.put(outputKey, outputValue);
+      }
+
+      return outputMap;
+    } else {
+      if (inputFieldType.getTypeName().equals(outputFieldType.getTypeName())) {
+        return inputValue;
+      } else {
+        KV<TypeName, TypeName> key =
+            KV.of(inputFieldType.getTypeName(), outputFieldType.getTypeName());
+        return Inference.ALL_MAP.get(key).apply(inputValue);
+      }
+    }
+  }
+
+  /** Configures casting for nullable fields. */
+  public enum Nullability {
+    /** Can cast only if nullability is the same. */
+    SAME,
+    /** Can cast non-nullable fields to nullable. */
+    WEAKEN,
+    /** Can cast nullable fields to non-nullable and forth. */
+    IGNORE
```

Comment:
The term `IGNORE` might be misleading. This would be a failed cast if the input is null. FWIW in this type system you can consider `NULLABLE t` to be a supertype of `t` so it is perfectly analogous to the `Type` enum.